pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,313 @@
+package logql
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/loki/pkg/logql/syntax"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+type RangeVectorMapper struct {
+	splitByInterval time.Duration
+}
+
+func NewRangeVectorMapper(interval time.Duration) (RangeVectorMapper, error) {
+	if interval <= 0 {
+		return RangeVectorMapper{}, fmt.Errorf("cannot create RangeVectorMapper with <=0 splitByInterval. Received %s", interval)
+	}
+	return RangeVectorMapper{
+		splitByInterval: interval,
+	}, nil
+}
+
+// Parse returns (noop, parsed expression, error)
+func (m RangeVectorMapper) Parse(query string) (bool, syntax.Expr, error) {
+	origExpr, err := syntax.ParseSampleExpr(query)
+	if err != nil {
+		return true, nil, err
+	}
+
+	if !isSplittableByRange(origExpr) {
+		return true, origExpr, nil
+	}
+
+	modExpr, err := m.Map(origExpr, nil)
+	if err != nil {
+		return true, nil, err
+	}
+
+	return origExpr.String() == modExpr.String(), modExpr, err
+}
+
+func (m RangeVectorMapper) Map(expr syntax.SampleExpr, vectorAggrPushdown *syntax.VectorAggregationExpr) (syntax.SampleExpr, error) {
+	// immediately clone the passed expr to avoid mutating the original
+	expr, err := clone(expr)
+	if err != nil {
+		return nil, err
+	}
+
+	switch e := expr.(type) {
+	case *syntax.VectorAggregationExpr:
+		return m.mapVectorAggregationExpr(e)
+	case *syntax.RangeAggregationExpr:
+		return m.mapRangeAggregationExpr(e, vectorAggrPushdown), nil
+	case *syntax.BinOpExpr:
+		lhsMapped, err := m.Map(e.SampleExpr, vectorAggrPushdown)
+		if err != nil {
+			return nil, err
+		}
+		rhsMapped, err := m.Map(e.RHS, vectorAggrPushdown)
+		if err != nil {
+			return nil, err
+		}
+		e.SampleExpr = lhsMapped
+		e.RHS = rhsMapped
+		return e, nil
+	default:
+		return nil, errors.Errorf("unexpected expr type (%T) for ASTMapper type (%T) ", expr, m)
+	}
+}
+
+// getRangeInterval returns the interval in the range vector
+// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
+func getRangeInterval(expr syntax.SampleExpr) time.Duration {
+	var rangeInterval time.Duration
+	expr.Walk(func(e interface{}) {
+		switch concrete := e.(type) {
+		case *syntax.RangeAggregationExpr:
+			rangeInterval = concrete.Left.Interval
+		}
+	})
+	return rangeInterval
+}
+
+// hasLabelExtractionStage returns true if an expression contains a stage for label extraction,
+// such as `| json` or `| logfmt`
+func hasLabelExtractionStage(expr syntax.SampleExpr) bool {
+	found := false
+	expr.Walk(func(e interface{}) {
+		switch concrete := e.(type) {
+		case *syntax.LabelParserExpr:
+			if concrete.Op == syntax.OpParserTypeJSON || concrete.Op == syntax.OpParserTypeLogfmt {
+				found = true
+			}
+		}
+	})
+	return found
+}
+
+// sumOverFullRange returns an expression that sum up individual downstream queries by preserving labels
```

Comment:
```suggestion
// sumOverFullRange returns an expression that sums up individual downstream queries by preserving labels
```

I'm not sure I understand `by preserving labels and dividing`. The example is clearer to me.

Reply:
Ok, will try to clarify the comment.

"sum by preserving labels" should mean, that it merges multiple results with a sum operation without dropping labels. In contrary to vector aggregations, such as `sum()` which return a single stream with a single value.

Reply:
So it's a sum by labels, isn't it?